corehq/motech/repeaters/tasks.py

```diff
@@ -72,65 +64,103 @@ def delete_old_request_logs():
     queue=settings.CELERY_PERIODIC_QUEUE,
 )
 def check_repeaters():
-    # this creates a task for all partitions
-    # the Nth child task determines if a lock is available for the Nth partition
-    for current_partition in range(CHECK_REPEATERS_PARTITION_COUNT):
-        check_repeaters_in_partition.delay(current_partition)
-
-
-@task(queue=settings.CELERY_PERIODIC_QUEUE)
-def check_repeaters_in_partition(partition):
-    """
-    The CHECK_REPEATERS_PARTITION_COUNT constant dictates the total number of partitions
-    :param partition: index of partition to check
-    """
     start = datetime.utcnow()
     twentythree_hours_sec = 23 * 60 * 60
     twentythree_hours_later = start + timedelta(hours=23)
 
-    # Long timeout to allow all waiting repeat records to be iterated
-    lock_key = f"{CHECK_REPEATERS_KEY}_{partition}_in_{CHECK_REPEATERS_PARTITION_COUNT}"
+    # Long timeout to allow all waiting repeaters to be iterated
+    # TODO: Check how long it takes to iterate all repeaters on prod
     check_repeater_lock = get_redis_lock(
-        lock_key,
+        CHECK_REPEATERS_KEY,
         timeout=twentythree_hours_sec,
-        name=lock_key,
+        name=CHECK_REPEATERS_KEY,
     )
     if not check_repeater_lock.acquire(blocking=False):
-        metrics_counter("commcare.repeaters.check.locked_out", tags={'partition': partition})
+        metrics_counter("commcare.repeaters.check.locked_out")
         return
 
     try:
         with metrics_histogram_timer(
             "commcare.repeaters.check.processing",
             timing_buckets=_check_repeaters_buckets,
         ):
-            for record in RepeatRecord.objects.iter_partition(
-                    start, partition, CHECK_REPEATERS_PARTITION_COUNT):
-                if not _soft_assert(
-                    datetime.utcnow() < twentythree_hours_later,
-                    "I've been iterating repeat records for 23 hours. I quit!"
-                ):
-                    break
-
+            for repeater in iter_ready_repeaters():
                 metrics_counter("commcare.repeaters.check.attempt_forward")
-                record.attempt_forward_now(is_retry=True)
-            else:
-                iterating_time = datetime.utcnow() - start
-                _soft_assert(
-                    iterating_time < timedelta(hours=6),
-                    f"It took {iterating_time} to iterate repeat records."
-                )
+                process_repeater.delay(repeater.domain, repeater.repeater_id)
```

Comment:
Is it dangerous to have two `process_repeater` tasks running concurrently for the same repeater? Is it true that this could enqueue a task to a queue that's backed up (by more than 5 minutes), and then 5 minutes later another `check_repeaters` call could do the same thing, and then two different workers pick up those tasks and execute them concurrently?

I think I'm not understanding how the lock will prevent concurrent tasks since it is only guarding when tasks are enqueued, not when they are run.

Reply:
I am imagining the following scenario: An environment has a repeater with more than `MAX_REPEATER_WORKERS` waiting repeat records, and the remote endpoint is timing out. Every time `iter_ready_repeaters()` loops through repeaters this repeater is returned, and `process_repeater()` sends off another `MAX_REPEATER_WORKERS` more requests. How many requests could we send before the first timeouts come back? :rofl: :scream: 

> how the lock will prevent concurrent tasks since it is only guarding when tasks are enqueued, not when they are run

Well this is exactly my issue with our broken Redis lock. Firstly, the lock should be acquired in `iter_ready_repeaters()` so that locked repeaters can't set `yielded` to `True`. And secondly, the lock must be released in the `update_repeater()` task, after processing is complete. Documentation suggests that this workflow is anticipated and supported, but `django-redis` (maybe?) is throwing an error that `update_repeater()` is trying to release a lock that isn't acquired, and the lock acquired in `iter_ready_repeaters()` is never released. :roll_eyes: